bndtools.core/src/bndtools/central/Central.java

```diff
@@ -212,22 +212,25 @@ public synchronized static Workspace getWorkspace() throws Exception {
             newWorkspace.addBasicPlugin(new WorkspaceListener(newWorkspace));
             newWorkspace.addBasicPlugin(instance.repoListenerTracker);
             newWorkspace.addBasicPlugin(getWorkspaceR5Repository());
+            newWorkspace.addBasicPlugin(new JobProgress());
 
             // Initialize projects in synchronized block
             newWorkspace.getBuildOrder();
 
             // Monitor changes in cnf so we can refresh the workspace
             addCnfChangeListener(newWorkspace);
 
-            // The workspace has been initialized fully, set the field now
-            workspace = newWorkspace;
-
-            workspaceRepositoryChangeDetector = new WorkspaceRepositoryChangeDetector(workspace);
+            workspaceRepositoryChangeDetector = new WorkspaceRepositoryChangeDetector(newWorkspace);
 
             // Call the queued workspace init callbacks
-            while (!workspaceInitCallbackQueue.isEmpty()) {
-                Function<Workspace,Void> callback = workspaceInitCallbackQueue.remove(0);
-                callback.run(workspace);
+            synchronized (workspaceInitCallbackQueue) {
```

Comment:
I am afraid of this new synchronization here. Especially since, while holding it, calls are made to the callbacks. I think it would be much safer to do something like is done in OSGi's PromiseImpl. Use a ConcurrentLinkedQueue to hold the call backs and a notifyCallbacks method which just returns if the workspace field is unset. This method can be called from both getWorkspace and onWorkspaceInit and it does not matter that multiple threads can be pulling callbacks off the queue and calling them. The Central class lock should probably not be be held either while calling the callbacks.

Just trying to avoid creating a future problem for ourselves :-) but I am open to this suggestion being overkill...